File: apple/TriageCompanion/AppModel.swift
import Foundation

@MainActor
final class AppModel: ObservableObject {
    @Published var draft = CredentialDraft()
    @Published var credentialStatuses: [ServiceCredentialStatus] = []
    @Published var settingsMessage: String?
    @Published var isSavingSettings = false

    @Published var notifications: LoadState<[GitHubNotificationItem]> = .idle
    @Published var dependabotAlerts: LoadState<[DependabotAlertItem]> = .idle
    @Published var failedWorkflows: LoadState<[FailedWorkflowRunItem]> = .idle
    @Published var snykIssues: LoadState<[SnykIssueItem]> = .idle
    @Published var jiraTickets: LoadState<[JiraTicketItem]> = .idle

    @Published var hasSavedGitHubToken = false
    @Published var hasSavedSnykToken = false
    @Published var hasSavedJiraToken = false

    private let store: CredentialStoring

    init(store: CredentialStoring = AppCredentialStore()) {
        self.store = store
        reloadCredentials()
    }

    func reloadCredentials() {
        do {
            let snapshot = try credentialSnapshot()
            hasSavedGitHubToken = snapshot.githubToken != nil
            hasSavedSnykToken = snapshot.snykToken != nil
            hasSavedJiraToken = snapshot.jiraAPIToken != nil
            draft.githubToken = ""
            draft.snykToken = ""
            draft.jiraAPIToken = ""
            draft.snykAPIBaseURL = snapshot.snykAPIBaseURL
            draft.jiraBaseURL = snapshot.jiraBaseURL ?? ""
            draft.jiraEmail = snapshot.jiraEmail ?? ""
            draft.jiraCloudID = snapshot.jiraCloudID ?? ""
            credentialStatuses = [
                ServiceCredentialStatus(
                    service: .github,
                    isConfigured: snapshot.hasGitHub,
                    detail: snapshot.hasGitHub ? "Token saved" : "Token required"
                ),
                ServiceCredentialStatus(
                    service: .snyk,
                    isConfigured: snapshot.hasSnyk,
                    detail: snapshot.hasSnyk ? "Token saved" : "Token required"
                ),
                ServiceCredentialStatus(
                    service: .jira,
                    isConfigured: snapshot.hasJira,
                    detail: snapshot.hasJira ? "Credentials saved" : "Base URL, email, and token required"
                )
            ]
        } catch {
            settingsMessage = errorMessage(error)
        }
    }

    func saveSettings() async {
        isSavingSettings = true
        defer { isSavingSettings = false }

        do {
            if !draft.githubToken.isEmpty {
                try store.save(try cleanRequiredText(draft.githubToken, label: "GitHub token"), for: .githubToken)
            }
            if !draft.snykToken.isEmpty {
                try store.save(try cleanRequiredText(draft.snykToken, label: "Snyk token"), for: .snykToken)
            }
            try store.save(try normalizedSnykAPIBaseURL(draft.snykAPIBaseURL), for: .snykAPIBaseURL)

            let hasJiraInput = !draft.jiraBaseURL.isEmpty || !draft.jiraEmail.isEmpty || !draft.jiraAPIToken.isEmpty || !draft.jiraCloudID.isEmpty
            if hasJiraInput {
                try store.save(try normalizedJiraBaseURL(draft.jiraBaseURL), for: .jiraBaseURL)
                try store.save(try cleanRequiredText(draft.jiraEmail, label: "Jira email"), for: .jiraEmail)
                if !draft.jiraAPIToken.isEmpty {
                    try store.save(try cleanRequiredText(draft.jiraAPIToken, label: "Jira API token"), for: .jiraAPIToken)
                } else if try store.read(.jiraAPIToken) == nil {
                    throw TriageCompanionError.message("Jira API token is required.")
                }
                try store.save(try validateCloudID(draft.jiraCloudID), for: .jiraCloudID)
            }

            settingsMessage = "Settings saved."
            reloadCredentials()
        } catch {
            settingsMessage = errorMessage(error)
        }
    }

    func removeGitHubToken() {
        do {
            try store.save(nil, for: .githubToken)
            settingsMessage = "GitHub token removed."
            reloadCredentials()
        } catch {
            settingsMessage = errorMessage(error)
        }
    }

    func removeSnykToken() {
        do {
            try store.save(nil, for: .snykToken)
            settingsMessage = "Snyk token removed."
            reloadCredentials()
        } catch {
            settingsMessage = errorMessage(error)
        }
    }

    func clearJiraCredentials() {
        do {
            try store.save(nil, for: .jiraBaseURL)
            try store.save(nil, for: .jiraEmail)
            try store.save(nil, for: .jiraAPIToken)
            try store.save(nil, for: .jiraCloudID)
            settingsMessage = "Jira credentials cleared."
            reloadCredentials()
        } catch {
            settingsMessage = errorMessage(error)
        }
    }

    func refreshAll() async {
        await refreshGitHub()
        await refreshSnyk()
        await refreshJira()
    }

    func refreshGitHub() async {
        let snapshot: CredentialSnapshot
        do {
            snapshot = try credentialSnapshot()
        } catch {
            let message = errorMessage(error)
            notifications = .failed(message)
            dependabotAlerts = .failed(message)
            failedWorkflows = .failed(message)
            return
        }

        guard let token = snapshot.githubToken else {
            let message = "Save a GitHub token in Settings."
            notifications = .notConfigured(message)
            dependabotAlerts = .notConfigured(message)
            failedWorkflows = .notConfigured(message)
            return
        }

        let client = GitHubClient(token: token)
        notifications = .loading
        dependabotAlerts = .loading
        failedWorkflows = .loading
        do {
            let checkedAt = Date()
            notifications = .loaded(try await client.listNotifications(), checkedAt)
            dependabotAlerts = .loaded(try await client.listDependabotAlertsFromNotifications(), Date())
            failedWorkflows = .loaded(try await client.listFailedWorkflowRunsFromNotifications(), Date())
        } catch {
            let message = errorMessage(error)
            if case .loading = notifications {
                notifications = .failed(message)
            }
            if case .loading = dependabotAlerts {
                dependabotAlerts = .failed(message)
            }
            if case .loading = failedWorkflows {
                failedWorkflows = .failed(message)
            }
        }
    }

    func refreshSnyk() async {
        let snapshot: CredentialSnapshot
        do {
            snapshot = try credentialSnapshot()
        } catch {
            snykIssues = .failed(errorMessage(error))
            return
        }

        guard let token = snapshot.snykToken else {
            snykIssues = .notConfigured("Save a Snyk token in Settings.")
            return
        }

        snykIssues = .loading
        do {
            let client = try SnykClient(token: token, baseURL: snapshot.snykAPIBaseURL)
            snykIssues = .loaded(try await client.listOpenIssues(), Date())
        } catch {
            snykIssues = .failed(errorMessage(error))
        }
    }

    func refreshJira() async {
        let snapshot: CredentialSnapshot
        do {
            snapshot = try credentialSnapshot()
        } catch {
            jiraTickets = .failed(errorMessage(error))
            return
        }

        guard let baseURL = snapshot.jiraBaseURL,
              let email = snapshot.jiraEmail,
              let apiToken = snapshot.jiraAPIToken else {
            jiraTickets = .notConfigured("Save Jira credentials in Settings.")
            return
        }

        jiraTickets = .loading
        do {
            let client = try JiraClient(
                baseURL: baseURL,
                email: email,
                apiToken: apiToken,
                cloudID: snapshot.jiraCloudID
            )
            jiraTickets = .loaded(try await client.listOpenTickets(), Date())
        } catch {
            jiraTickets = .failed(errorMessage(error))
        }
    }

    private func credentialSnapshot() throws -> CredentialSnapshot {
        let snykAPIBaseURL = try normalizedSnykAPIBaseURL(try store.read(.snykAPIBaseURL) ?? defaultSnykAPIBaseURL)
        let cloudID = try validateCloudID(try store.read(.jiraCloudID) ?? "")

        return CredentialSnapshot(
            githubToken: try storedRequiredText(.githubToken, label: "GitHub token"),
            snykToken: try storedRequiredText(.snykToken, label: "Snyk token"),
            snykAPIBaseURL: snykAPIBaseURL,
            jiraBaseURL: try storedNormalizedJiraBaseURL(),
            jiraEmail: try storedRequiredText(.jiraEmail, label: "Jira email"),
            jiraAPIToken: try storedRequiredText(.jiraAPIToken, label: "Jira API token"),
            jiraCloudID: cloudID
        )
    }

    private func storedRequiredText(_ key: CredentialKey, label: String) throws -> String? {
        guard let value = try store.read(key) else {
            return nil
        }

        return try cleanRequiredText(value, label: label)
    }

    private func storedNormalizedJiraBaseURL() throws -> String? {
        guard let value = try store.read(.jiraBaseURL) else {
            return nil
        }

        return try normalizedJiraBaseURL(value)
    }

    private func errorMessage(_ error: Error) -> String {
        if let localized = error as? LocalizedError, let description = localized.errorDescription {
            return description
        }

        return error.localizedDescription
    }
}

File: android/app/src/main/kotlin/com/triagecompanion/app/AppModel.kt
package com.triagecompanion.app

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import java.time.Instant
import kotlinx.coroutines.launch

class CredentialDraft {
    var githubToken by mutableStateOf("")
    var snykToken by mutableStateOf("")
    var snykAPIBaseURL by mutableStateOf(DEFAULT_SNYK_API_BASE_URL)
    var jiraBaseURL by mutableStateOf("")
    var jiraEmail by mutableStateOf("")
    var jiraAPIToken by mutableStateOf("")
    var jiraCloudID by mutableStateOf("")
}

class AppModel(application: Application) : AndroidViewModel(application) {
    private val store: CredentialStoring = AppCredentialStore(application)

    val draft = CredentialDraft()
    var credentialStatuses by mutableStateOf(listOf<ServiceCredentialStatus>())
        private set
    var settingsMessage by mutableStateOf<String?>(null)
        private set
    var isSavingSettings by mutableStateOf(false)
        private set

    var notifications by mutableStateOf<LoadState<List<GitHubNotificationItem>>>(LoadState.Idle)
        private set
    var dependabotAlerts by mutableStateOf<LoadState<List<DependabotAlertItem>>>(LoadState.Idle)
        private set
    var failedWorkflows by mutableStateOf<LoadState<List<FailedWorkflowRunItem>>>(LoadState.Idle)
        private set
    var snykIssues by mutableStateOf<LoadState<List<SnykIssueItem>>>(LoadState.Idle)
        private set
    var jiraTickets by mutableStateOf<LoadState<List<JiraTicketItem>>>(LoadState.Idle)
        private set

    var hasSavedGitHubToken by mutableStateOf(false)
        private set
    var hasSavedSnykToken by mutableStateOf(false)
        private set
    var hasSavedJiraToken by mutableStateOf(false)
        private set

    init {
        reloadCredentials()
    }

    fun reloadCredentials() {
        try {
            val snapshot = credentialSnapshot()
            hasSavedGitHubToken = snapshot.githubToken != null
            hasSavedSnykToken = snapshot.snykToken != null
            hasSavedJiraToken = snapshot.jiraAPIToken != null
            draft.githubToken = ""
            draft.snykToken = ""
            draft.jiraAPIToken = ""
            draft.snykAPIBaseURL = snapshot.snykAPIBaseURL
            draft.jiraBaseURL = snapshot.jiraBaseURL ?: ""
            draft.jiraEmail = snapshot.jiraEmail ?: ""
            draft.jiraCloudID = snapshot.jiraCloudID ?: ""
            credentialStatuses = listOf(
                ServiceCredentialStatus(
                    service = ServiceId.GITHUB,
                    isConfigured = snapshot.hasGitHub,
                    detail = if (snapshot.hasGitHub) "Token saved" else "Token required",
                ),
                ServiceCredentialStatus(
                    service = ServiceId.SNYK,
                    isConfigured = snapshot.hasSnyk,
                    detail = if (snapshot.hasSnyk) "Token saved" else "Token required",
                ),
                ServiceCredentialStatus(
                    service = ServiceId.JIRA,
                    isConfigured = snapshot.hasJira,
                    detail = if (snapshot.hasJira) "Credentials saved" else "Base URL, email, and token required",
                ),
            )
        } catch (error: Exception) {
            settingsMessage = errorMessage(error)
        }
    }

    fun saveSettings() {
        viewModelScope.launch {
            isSavingSettings = true
            try {
                if (draft.githubToken.isNotEmpty()) {
                    store.save(cleanRequiredText(draft.githubToken, label = "GitHub token"), CredentialKey.GITHUB_TOKEN)
                }
                if (draft.snykToken.isNotEmpty()) {
                    store.save(cleanRequiredText(draft.snykToken, label = "Snyk token"), CredentialKey.SNYK_TOKEN)
                }
                store.save(normalizedSnykAPIBaseURL(draft.snykAPIBaseURL), CredentialKey.SNYK_API_BASE_URL)

                val hasJiraInput = draft.jiraBaseURL.isNotEmpty() || draft.jiraEmail.isNotEmpty() ||
                    draft.jiraAPIToken.isNotEmpty() || draft.jiraCloudID.isNotEmpty()
                if (hasJiraInput) {
                    store.save(normalizedJiraBaseURL(draft.jiraBaseURL), CredentialKey.JIRA_BASE_URL)
                    store.save(cleanRequiredText(draft.jiraEmail, label = "Jira email"), CredentialKey.JIRA_EMAIL)
                    if (draft.jiraAPIToken.isNotEmpty()) {
                        store.save(
                            cleanRequiredText(draft.jiraAPIToken, label = "Jira API token"),
                            CredentialKey.JIRA_API_TOKEN,
                        )
                    } else if (store.read(CredentialKey.JIRA_API_TOKEN) == null) {
                        throw TriageCompanionException("Jira API token is required.")
                    }
                    store.save(validateCloudID(draft.jiraCloudID), CredentialKey.JIRA_CLOUD_ID)
                }

                settingsMessage = "Settings saved."
                reloadCredentials()
            } catch (error: Exception) {
                settingsMessage = errorMessage(error)
            } finally {
                isSavingSettings = false
            }
        }
    }

    fun removeGitHubToken() {
        try {
            store.save(null, CredentialKey.GITHUB_TOKEN)
            settingsMessage = "GitHub token removed."
            reloadCredentials()
        } catch (error: Exception) {
            settingsMessage = errorMessage(error)
        }
    }

    fun removeSnykToken() {
        try {
            store.save(null, CredentialKey.SNYK_TOKEN)
            settingsMessage = "Snyk token removed."
            reloadCredentials()
        } catch (error: Exception) {
            settingsMessage = errorMessage(error)
        }
    }

    fun clearJiraCredentials() {
        try {
            store.save(null, CredentialKey.JIRA_BASE_URL)
            store.save(null, CredentialKey.JIRA_EMAIL)
            store.save(null, CredentialKey.JIRA_API_TOKEN)
            store.save(null, CredentialKey.JIRA_CLOUD_ID)
            settingsMessage = "Jira credentials cleared."
            reloadCredentials()
        } catch (error: Exception) {
            settingsMessage = errorMessage(error)
        }
    }

    fun refreshAll() {
        viewModelScope.launch {
            refreshGitHubNow()
            refreshSnykNow()
            refreshJiraNow()
        }
    }

    fun refreshGitHub() {
        viewModelScope.launch { refreshGitHubNow() }
    }

    fun refreshSnyk() {
        viewModelScope.launch { refreshSnykNow() }
    }

    fun refreshJira() {
        viewModelScope.launch { refreshJiraNow() }
    }

    private suspend fun refreshGitHubNow() {
        val snapshot = try {
            credentialSnapshot()
        } catch (error: Exception) {
            val message = errorMessage(error)
            notifications = LoadState.Failed(message)
            dependabotAlerts = LoadState.Failed(message)
            failedWorkflows = LoadState.Failed(message)
            return
        }

        val token = snapshot.githubToken
        if (token == null) {
            val message = "Save a GitHub token in Settings."
            notifications = LoadState.NotConfigured(message)
            dependabotAlerts = LoadState.NotConfigured(message)
            failedWorkflows = LoadState.NotConfigured(message)
            return
        }

        val client = GitHubClient(token)
        notifications = LoadState.Loading
        dependabotAlerts = LoadState.Loading
        failedWorkflows = LoadState.Loading
        try {
            notifications = LoadState.Loaded(client.listNotifications(), Instant.now())
            dependabotAlerts = LoadState.Loaded(client.listDependabotAlertsFromNotifications(), Instant.now())
            failedWorkflows = LoadState.Loaded(client.listFailedWorkflowRunsFromNotifications(), Instant.now())
        } catch (error: Exception) {
            val message = errorMessage(error)
            if (notifications == LoadState.Loading) {
                notifications = LoadState.Failed(message)
            }
            if (dependabotAlerts == LoadState.Loading) {
                dependabotAlerts = LoadState.Failed(message)
            }
            if (failedWorkflows == LoadState.Loading) {
                failedWorkflows = LoadState.Failed(message)
            }
        }
    }

    private suspend fun refreshSnykNow() {
        val snapshot = try {
            credentialSnapshot()
        } catch (error: Exception) {
            snykIssues = LoadState.Failed(errorMessage(error))
            return
        }

        val token = snapshot.snykToken
        if (token == null) {
            snykIssues = LoadState.NotConfigured("Save a Snyk token in Settings.")
            return
        }

        snykIssues = LoadState.Loading
        try {
            val client = SnykClient(token, baseURL = snapshot.snykAPIBaseURL)
            snykIssues = LoadState.Loaded(client.listOpenIssues(), Instant.now())
        } catch (error: Exception) {
            snykIssues = LoadState.Failed(errorMessage(error))
        }
    }

    private suspend fun refreshJiraNow() {
        val snapshot = try {
            credentialSnapshot()
        } catch (error: Exception) {
            jiraTickets = LoadState.Failed(errorMessage(error))
            return
        }

        val baseURL = snapshot.jiraBaseURL
        val email = snapshot.jiraEmail
        val apiToken = snapshot.jiraAPIToken
        if (baseURL == null || email == null || apiToken == null) {
            jiraTickets = LoadState.NotConfigured("Save Jira credentials in Settings.")
            return
        }

        jiraTickets = LoadState.Loading
        try {
            val client = JiraClient(baseURL, email, apiToken, cloudID = snapshot.jiraCloudID)
            jiraTickets = LoadState.Loaded(client.listOpenTickets(), Instant.now())
        } catch (error: Exception) {
            jiraTickets = LoadState.Failed(errorMessage(error))
        }
    }

    private fun credentialSnapshot(): CredentialSnapshot {
        val snykAPIBaseURL = normalizedSnykAPIBaseURL(
            store.read(CredentialKey.SNYK_API_BASE_URL) ?: DEFAULT_SNYK_API_BASE_URL,
        )
        val cloudID = validateCloudID(store.read(CredentialKey.JIRA_CLOUD_ID) ?: "")

        return CredentialSnapshot(
            githubToken = storedRequiredText(CredentialKey.GITHUB_TOKEN, label = "GitHub token"),
            snykToken = storedRequiredText(CredentialKey.SNYK_TOKEN, label = "Snyk token"),
            snykAPIBaseURL = snykAPIBaseURL,
            jiraBaseURL = storedNormalizedJiraBaseURL(),
            jiraEmail = storedRequiredText(CredentialKey.JIRA_EMAIL, label = "Jira email"),
            jiraAPIToken = storedRequiredText(CredentialKey.JIRA_API_TOKEN, label = "Jira API token"),
            jiraCloudID = cloudID,
        )
    }

    private fun storedRequiredText(key: CredentialKey, label: String): String? {
        val value = store.read(key) ?: return null

        return cleanRequiredText(value, label)
    }

    private fun storedNormalizedJiraBaseURL(): String? {
        val value = store.read(CredentialKey.JIRA_BASE_URL) ?: return null

        return normalizedJiraBaseURL(value)
    }

    private fun errorMessage(error: Exception): String =
        error.message ?: error.javaClass.simpleName
}
